src/la/arnoldi/dbcsr_arnoldi_api.F

```diff
@@ -0,0 +1,545 @@
+            CALL gev_arnoldi_init(matrix, matrix_arnoldi, vectors, arnoldi_data)
+         END IF
+
+! Generate the subspace
+         CALL gev_build_subspace(matrix_arnoldi, vectors, arnoldi_data)
+
+! If we reached the maximum number of steps or the subspace converged we still need to get the eigenvalues
+         CALL compute_evals(arnoldi_data)
+
+! Select the evals according to user selection and keep them in arnoldi_data
+         CALL select_evals(arnoldi_data)
+
+! update the matrices and compute the convergence
+         CALL gev_update_data(matrix, matrix_arnoldi, vectors, arnoldi_data)
+
+! Check whether we can already go home
+         IF (control%converged) EXIT
+      END DO
+
+! Deallocated the work vectors
+      CALL dbcsr_release(vectors%input_vec)
+      CALL dbcsr_release(vectors%result_vec)
+      CALL dbcsr_release(vectors%rep_col_vec)
+      CALL dbcsr_release(vectors%rep_row_vec)
+      CALL dbcsr_release(A_rho_B)
+      DEALLOCATE (matrix_arnoldi)
+
+      CALL timestop(handle)
+
+   END SUBROUTINE arnoldi_generalized_ev
+
+! **************************************************************************************************
+!> \brief simple wrapper to estimate extremal eigenvalues with arnoldi, using the old lanczos interface
+!>        this hides some of the power of the arnoldi routines (e.g. only min or max eval, generalized eval, ritz vectors, etc.),
+!>        and does not allow for providing an initial guess of the ritz vector.
+!> \param matrix_a input mat
+!> \param max_ev estimated max eval
+!> \param min_ev estimated min eval
+!> \param converged Usually arnoldi is more accurate than claimed.
+!> \param threshold target precision
+!> \param max_iter max allowed iterations (will be rounded up)
+! **************************************************************************************************
+   SUBROUTINE arnoldi_extremal(matrix_a, max_ev, min_ev, converged, threshold, max_iter)
+      TYPE(dbcsr_type), INTENT(INOUT), TARGET            :: matrix_a
+      REAL(KIND=dp), INTENT(OUT)                         :: max_ev, min_ev
+      LOGICAL, INTENT(OUT)                               :: converged
+      REAL(KIND=dp), INTENT(IN)                          :: threshold
+      INTEGER, INTENT(IN)                                :: max_iter
+
+      CHARACTER(LEN=*), PARAMETER :: routineN = 'arnoldi_extremal', &
+                                     routineP = moduleN//':'//routineN
+
+      INTEGER                                            :: handle, max_iter_internal, nrestarts
+      TYPE(arnoldi_data_type)                            :: my_arnoldi
+      TYPE(dbcsr_p_type), DIMENSION(:), POINTER          :: arnoldi_matrices
+
+      CALL timeset(routineN, handle)
+
+      ! we go in chunks of max_iter_internal, and restart ater each of those.
+      ! at low threshold smaller values of max_iter_internal make sense
+      IF (.TRUE.) max_iter_internal = 16
+      IF (threshold <= 1.0E-3_dp) max_iter_internal = 32
```

Comment:
magic values?

Reply:
there are several empirical values. 16 32 64 1.0E-3_dp  1.0E-4_dp. I think it does not make too much sense to create for all of them a constant variable. The function is small itself and everything is clear.